### ISSUE TITLE
[HOLD for payment 2023-11-13] [$500] Task needs revisit to update members

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#29228]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.80-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1696942077360429

## Action Performed:
1. Open the app
2. Open any task report or raise one and open the report
3. Click on task header -> members to observe current members
4. Again get back to report and change assignee to any non member user
5. Again click on task header -> members and observe that new assignee is not added to members
6. Revisit report and observe members, it will now display new assignee in members
## Expected Result: 
App should add new assignee to members as soon as we change the assignee
## Actual Result: 
App needs revisit to report to add new assignee to members
## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/38435837/5cc61d7f-eab6-4111-83b5-afa24cead3fb

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/38435837/25df0787-2078-48b9-98b6-449203fd8611

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/38435837/cbbe4c53-c879-4986-9bb8-7b03287b08ed

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/38435837/29b75e3d-b0eb-4991-be2f-f02379adec15

https://github.com/Expensify/App/assets/38435837/5f865898-14d7-4c5f-9f25-1db2ca0a3728

https://github.com/Expensify/App/assets/38435837/f145594c-2cbc-49e0-9ba4-23309aaa2d31

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/38435837/ba719838-c6f0-486e-ad9e-cb9ddbdfafe7

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01609c654d88438959</li>
        <li>Upwork Job ID: 1711835475398246400</li>
        <li>Last Price Increase: 2023-10-24</li>
<li>Automatic offers: </li>
<ul>
<li>cubuspl42 | Reviewer | 27439450</li>
<li>Tony-MK | Contributor | 27439452</li>
<li>dhanashree-sawant | Reporter | 27439455</li>
</ul></li>
    </ul>
</details>

**Comment by stephanieelliott** [comment]
> I'm not able to reproduce this issue.
In step 3, it shows profile page instead of task details page when the user clicks the header of task.
Is this bug or expected behavior?

I'm seeing the same thing as @akamefi202. @dhanashree-sawant can you expand on the repro steps? Or are there specific settings on the task that needs to be set in order to repro this?

**Comment by dhanashree-sawant** [comment]
I saw the same issue when creating new task i.e. it displayed user profile (seems to be different bug). What I did was I changed the assignee to someone else and revisited. After that it started displaying members and settings option in details page.

**Comment by stephanieelliott** [comment]
still waiting for proposals

**Comment by Tony-MK** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Task needs revisit to update members.

### What is the root cause of that problem?

The [`ReportParticipantsPage`](https://github.com/Expensify/App/blob/main/src/pages/ReportParticipantsPage.js) uses the `report.participantAccountIDs` function to get the member list from [`ReportUtils.getParticipantsIDs`](https://github.com/Expensify/App/blob/a59d56f00a67607a1dbae1b422807f76f548f1f0/src/libs/ReportUtils.js#L4028).

https://github.com/Expensify/App/blob/a59d56f00a67607a1dbae1b422807f76f548f1f0/src/libs/ReportUtils.js#L4028-L4033

https://github.com/Expensify/App/blob/a59d56f00a67607a1dbae1b422807f76f548f1f0/src/pages/ReportParticipantsPage.js#L57-L58


The `report.participantAccountIDs` is updated when the report data is retrieved from the `command=openReport` API response. This happens when the user opens a report.

https://github.com/Expensify/App/blob/a59d56f00a67607a1dbae1b422807f76f548f1f0/src/libs/actions/Task.js#L432

Hence, the root cause of the problem is that when the user changes the task assignee, the [`Task.editTaskAssigneeAndNavigate`](https://github.com/Expensify/App/blob/main/src/libs/actions/Task.js#L432) function assumes that `command=openReport` will be called soon to update `report.participantAccountIDs`. However, this is not the case. Unless the user closes and revisits the task report, then the `command=openReport` request will be processed, hence updating `report.participantAccountIDs`. Therefore, the [`ReportParticipantsPage`](https://github.com/Expensify/App/blob/main/src/pages/ReportParticipantsPage.js) can list the recently added members in realtime.

### What changes do you think we should make in order to solve the problem?

In order to solve the problem, I suggest updating the `report.participantAccountIDs` array instead of waiting on `command=openReport`. Hence, we add `participantAccountIDs : [...report.participantAccountIDs, assigneeAccountID]` in the **second** item in the `optimisticData` array, but only when the `report.participantAccountIDs` does not contain the `assigneeAccountID`. This will allow the new assignee to be updated in the [`ReportParticipantsPage`](https://github.com/Expensify/App/blob/main/src/pages/ReportParticipantsPage.js) before the server updates the task report data. This will also improve the offline UX because the member's list will be updated without the need to wait for the report data in the Onyx to be updated with the `command=openReport` API response.

https://github.com/Expensify/App/blob/a59d56f00a67607a1dbae1b422807f76f548f1f0/src/libs/actions/Task.js#L481

Therefore, I believe adding something similar to the code below, after the condition above in [`Task.editTaskAssigneeAndNavigate`](https://github.com/Expensify/App/blob/main/src/libs/actions/Task.js#L432), should solve the problem.

```
if (report.participantAccountIDs && report.participantAccountIDs.include(assigneeAccountID)) optimisticData[1].value.participantAccountIDs = [...report.participantAccountIDs, assigneeAccountID];
```

**Comment by cubuspl42** [comment]
@Tony-MK 

> will be updated . very

Are we missing a sentence here, or is it just an extra word left in the proposal by accident?

**Comment by Tony-MK** [comment]
> @Tony-MK 
> 
> > will be updated . very
> 
> Are we missing a sentence here, or is it just an extra word left in the proposal by accident?

@cubuspl42 Sorry about that. It was an extra word left out while trying to edit the proposal.

**Comment by cubuspl42** [comment]
In my opinion, the proposal by @Tony-MK sounds reasonable. I agree that the suggested change will also improve the offline experience. I think we should proceed with @Tony-MK.

C+ reviewed 🎀 👀 🎀

**Comment by cubuspl42** [comment]
@luacmartins Do you know any reasons _not to_ update the `participantAccountIDs` optimistically in `editTaskAssigneeAndNavigate`?

**Comment by luacmartins** [comment]
I think we should be good to set `participantAccountIDs` optimistically in `editTaskAssigneeAndNavigate `. The only edge case I can think of is if the new assignee is an unknown user. In that case the accountID would be optimistic and might differ from the real accountID, but in that case the server should be updating that assignee accountID.

**Comment by stephanieelliott** [comment]
Disregard Melvin message above, this one is awaiting PR from @Tony-MK

**Comment by Tony-MK** [comment]
Hey @cubuspl42, I have created the [pull request](https://github.com/Expensify/App/pull/30671). Could you kindly find some time to review it?

**Comment by Tony-MK** [comment]
Hey @cubuspl42, I have created a [PR](https://github.com/Expensify/App/pull/30671). Could you kindly find some time to review it?
Thank you 😃

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by stephanieelliott** [comment]
PR deployed to staging on 11/2

**Comment by stephanieelliott** [comment]
Hey @cubuspl42 can you complete the BZ checklist please?

**Comment by cubuspl42** [comment]
- The PR that introduced the bug has been identified. Link to the PR:
    -  It seems that the bug was present in the relevant components since they were created
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
    - N/A
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
    - No need for additional discussion
- Determine if we should create a regression test for this bug.
    - Up to the QA team
- If we decide to create a regression test for the bug, please propose the regression test steps to ensure the same bug will not reach production again.
    - Create a task.
    - Assign the task to an account other than your own.
    - Click on the task header.
    - Confirm that the assignee is listed.
    - Reassign the task to a yet another account (not yours or the one previously selected).
    - Click the task report header again.
    - Ensure that the member count has incremented by one.
    - Confirm that the members list now includes the newly assigned member.

**Comment by stephanieelliott** [comment]
Summarizing payment on this issue: 

 - Issue reporter: @dhanashree-sawant $50 via Upwork -- PAID
 - Contributor: @Tony-MK $500 via Upwork -- PAID
 - Contributor+: @cubuspl42 $500via Upwork  -- PAID
 
 Upwork job is here, no bonus is included on payout
